django/db/backends/mysql/schema.py

```diff
@@ -135,24 +153,63 @@ def _field_should_be_indexed(self, model, field):
             return False
         return not self._is_limited_data_type(field)
 
-    def _delete_composed_index(self, model, fields, *args):
+    def _create_missing_fk_index(
+        self,
+        model,
+        *,
+        fields,
+        expressions=None,
+    ):
         """
         MySQL can remove an implicit FK index on a field when that field is
         covered by another index like a unique_together. "covered" here means
         that the more complex index starts like the simpler one.
-        https://bugs.mysql.com/bug.php?id=37910 / Django ticket #24757
-        We check here before removing the [unique|index]_together if we have to
+        https://bugs.mysql.com/bug.php?id=37910 / Django tickets #24757,31335
+        We check here before removing the composed index if we have to
         recreate a FK index.
         """
-        first_field = model._meta.get_field(fields[0])
+
+        first_field_name = None
+        if fields:
+            if (
+                fields[0].startswith("-")
+                and not self.connection.features.supports_index_column_ordering
+            ):
+                first_field_name = fields[0][1:]
+            elif not fields[0].startswith("-"):
+                first_field_name = fields[0]
```

Comment:
This is required only for indexes defined in `Meta.indexes`. Should we pass list of pure fields names in `remove_index()` instead? For example:
```python
    def remove_index(self, model, index):
        self._create_missing_fk_index(
            model,
            fields=[field_name for field_name, _ in index.fields_orders],
            expressions=index.expressions,
        )
        super().remove_index(model, index)
```